### PR TITLE
feat(session-replay-browser): merge consecutive mutation incremental snapshots

### PR DESCRIPTION
**Pull request by lewgordon-amplitude**
## Summary

- Adds `mergeMutationEvents` utility that coalesces consecutive rrweb `IncrementalSnapshot` mutation events into a single equivalent event
- Wires the merge into `EventCompressor.processQueue` and the full-snapshot drain path
- `isAttachIframe` events are never merged — they carry a full iframe document tree and must remain isolated
- Opt-in via `performanceConfig.mergeMutations` (defaults to `false`); existing customers are unaffected until they enable it

Resolves [SR-3752](https://linear.app/amplitude/issue/SR-3752)

## How merging is safe

Verified against the rrweb source at `~/repos/rrweb`:

| Field | Merge rule | Why safe |
|---|---|---|
| `removes` | concatenate in event order | replayer always applies removes before adds; IDs never reused within a session |
| `adds` | concatenate in event order | replayer's `queue`/retry loop already handles forward `nextId` references within a single event's adds list |
| `texts` | concatenate in event order | replayer calls `uniqueTextMutations()` which deduplicates keeping last-per-node-ID |
| `attributes` | concatenate in event order | last write per `(nodeId, attributeName)` wins — correct final state |
| `isAttachIframe` | never merged | these events deliver a whole iframe document tree and must stay isolated |

## Risk

**The main risk is node moves across event boundaries.** rrweb handles intra-callback moves via `movedMap` (remove from old parent + add to new parent in same event), but cross-event moves appear as a plain remove in one event and an add in another. When merged, both land in the same event — the replayer handles this correctly, but if the moved node's `nextId` also moved in the same batch, sibling ordering could resolve incorrectly. This isn't covered by unit tests and would only surface through manual replay validation.

**Timestamp precision:** merged events use the timestamp of the first event in the group. Intermediate DOM states within the merge window appear to happen at one instant, slightly degrading scrubbing precision during high-mutation periods.

### Unit tests (26 tests, `test/merge-mutation-events.test.ts`)

- Basic merge: single event passthrough, two-event merge, three-event merge, empty array
- Non-mutation events: `isAttachIframe` isolation, non-mutation `IncrementalSnapshot` passthrough, `FullSnapshot` passthrough
- Mixed sequences: mutations separated by scroll events, multi-run sequences
- Transient node elision: add+remove cancelled, cascade to orphaned children, texts/attributes filtered for elided nodes
- DOM move correctness: remove-then-re-add (cross-event move) preserved, add-then-move preserved at new parent, pre-existing node removed/re-added/removed not misidentified as transient, child moved away from transient parent survives at final parent

### E2E tests (4 tests, `e2e/mutation-merge.spec.ts`)

- Synchronous mutations captured and delivered to the API
- Transient node (add+remove in same MO batch) elided by rrweb — no add or remove in output
- Pre-existing node removed/re-added/removed: re-add cancelled, original removal preserved
- DOM move (remove from one parent, add under another): both operations preserved

### Automated review (Cursor Bugbot — 5 rounds)

Issues found and fixed iteratively before opening for human review:

| Round | Severity | Issue | Fix |
|---|---|---|---|
| 1 | Medium | Naive merge puts removes before adds; add-then-remove leaves stale node in DOM | Transient node elision: cancel both the add and remove for nodes created and removed within the window |
| 2a | Medium | Order-blind transient detection wrongly elides DOM moves (remove-then-re-add) | Track `lastAddEventIndex` + `lastRemoveEventIndex`; only transient when `addIdx < removeIdx` |
| 2b | Low | `texts` and `attributes` not filtered for transient node IDs | Filter all four arrays by the transient set |
| 3a | Medium | `firstAddEventIndex` breaks add-then-move (add in event0, move in event1 sets `lastRemoveIdx` = 1 = `lastAddIdx`) | Switch to `lastAddEventIndex` (last-write-wins) so the post-move re-add index equals the remove index |
| 3b | Low | Re-merging already-merged `taskQueue` entries on subsequent idle callbacks corrupts move semantics | Split into `pendingQueue` (new arrivals) and `taskQueue` (already merged); only merge the pending batch each pass |
| 4 | Medium | Cascade used `allAdds` (historical parentIds including pre-move positions); child moved from transient parent to stable parent wrongly elided | Track `lastParentById` (final parent only); cascade iterates that map instead |
| 5 | Medium | Pre-existing node removed in event1, re-added in event2, removed in event3: `lastAddIdx(1) < lastRemoveIdx(2)` falsely marks it transient, dropping the event1 remove | Also track `firstAddEventIndex`/`firstRemoveEventIndex`; require `firstAddIdx < firstRemoveIdx` as a second gate |

### Remaining manual validation
- [ ] Manual replay on a React-heavy app — confirm no "node not found" errors in the player
- [ ] Monitor event count reduction in staging

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Opt-in change to how rrweb mutation events are buffered and emitted, which can subtly affect replay correctness/timing (especially around DOM moves) despite added test coverage.
> 
> **Overview**
> Adds an opt-in `performanceConfig.mergeMutations` flag that merges consecutive rrweb mutation `IncrementalSnapshot` events into a single equivalent event before compression.
> 
> Updates `EventCompressor` to buffer events in a new `pendingQueue`, merge pending mutation runs (per `sessionId`) during idle processing and when draining prior to `FullSnapshot`, and avoid re-merging already-merged tasks.
> 
> Introduces `mergeMutationEvents` with logic to preserve replay semantics by filtering transient add/remove cycles (including cascades) and skipping `isAttachIframe` mutations, plus adds unit and Playwright E2E coverage and a test harness toggle (`sr-capture-test.html?mergeMutations=true`).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 412093587fe9a20b379fd4cffcf9ff05dd55ff10. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+function mergeGroup(events: eventWithTime[]): eventWithTime {
+  const first = events[0];
+  const merged: mutationData = {
+    source: IncrementalSource.Mutation,
+    removes: events.flatMap((e) => (e.data as mutationData).removes),
+    adds: events.flatMap((e) => (e.data as mutationData).adds),
+    texts: events.flatMap((e) => (e.data as mutationData).texts),
+    attributes: events.flatMap((e) => (e.data as mutationData).attributes),
+  };
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts:159]: Fixed in 9d4b3db. The root cause is exactly as described: when node X is added in event1 and removed in event2, the naive merge puts all removes before all adds, so the remove fires before X exists — gets skipped — then X is added and stays in the DOM permanently.

Fix: after concatenating all adds and removes from the merge window, detect transient nodes (any node whose id appears in both the adds and removes sets) and cancel both the add and the remove. We also cascade the cancellation to any adds whose parentId is a transient node, so that orphaned children of a transient parent don't get incorrectly inserted.

Added 4 new tests covering: direct add+remove elision, pre-existing removes preserved, cascade to children, and partial elision in a mixed event.

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  // Elide transient nodes: a node added in an earlier event and removed in a later one
+  // would otherwise have its remove execute before its add (since we put all removes
+  // first), leaving the node permanently in the DOM. Cancel both the add and the remove,
+  // and cascade to any adds whose parent is also transient.
+  const removedIds = new Set(allRemoves.map((r) => r.id));
+  const transientIds = new Set<number>();
+  for (const add of allAdds) {
+    if (removedIds.has(add.node.id)) transientIds.add(add.node.id);
+  }
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts:63]: Fixed in f18ac892. The previous set-intersection check was indeed order-blind. Fix: instead of building flat addedIds/removedIds sets, we now track `firstAddEventIndex` (first event index a node is added) and `lastRemoveEventIndex` (last event index a node is removed) by iterating through events in order. A node is only marked transient when `addIdx < removeIdx` — meaning it was added *before* it was removed. A DOM move (remove in event1, re-add in event2) has `addIdx > removeIdx` and is correctly preserved. Added a regression test for this case.

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      }
+    }
+  }
+
+  const merged: mutationData = {
+    source: IncrementalSource.Mutation,
+    removes: transientIds.size > 0 ? allRemoves.filter((r) => !transientIds.has(r.id)) : allRemoves,
+    adds: transientIds.size > 0 ? allAdds.filter((a) => !transientIds.has(a.node.id)) : allAdds,
+    texts: events.flatMap((e) => (e.data as mutationData).texts),
+    attributes: events.flatMap((e) => (e.data as mutationData).attributes),
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts]: Fixed in f18ac892. After computing transientIds, we now also filter `texts` and `attributes` arrays to drop any entries whose `id` is in transientIds. Added tests for both: text mutations on a transient node are dropped, and attribute mutations on a transient node are dropped.

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    for (const remove of data.removes) {
+      lastRemoveEventIndex.set(remove.id, i);
+    }
+  });
+
+  const transientIds = new Set<number>();
+  for (const [id, addIdx] of firstAddEventIndex) {
+    const removeIdx = lastRemoveEventIndex.get(id);
+    if (removeIdx !== undefined && addIdx < removeIdx) transientIds.add(id);
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts]: Fixed in 6daa34d. The root cause is using firstAddEventIndex: for add-then-move (add in event0, remove+re-add in event1), firstAddIdx=0 < lastRemoveIdx=1 → wrongly transient. Fix: switched to lastAddEventIndex (always overwrite). For add-then-move, the re-add in event1 sets lastAddIdx=1, which equals lastRemoveIdx=1, so lastAddIdx < lastRemoveIdx is false → not transient. The three cases now all work correctly: true transient (add0, remove1: lastAddIdx=0 < 1 ✓), DOM move (remove0, add1: lastAddIdx=1 > 0 → false ✓), add-then-move (add0, remove1+add1: lastAddIdx=1 = 1 → false ✓). Added a regression test.

File: packages/session-replay-browser/src/events/event-compressor.ts [hunk 5]
```diff
@@ -113,6 +114,7 @@ export class EventCompressor {
 
   // Process the task queue during idle time
   public processQueue(idleDeadline: IdleDeadline): void {
+    this.taskQueue = this.mergeMutationTasks(this.taskQueue);
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/event-compressor.ts]: Fixed in 6daa34d. Split the single taskQueue into pendingQueue (newly arrived events) and taskQueue (already merged, ready to process). enqueueEvent now pushes to pendingQueue. processQueue merges only pendingQueue and appends results to taskQueue, so already-merged events are never fed back through the transient detection logic. The full-snapshot drain path similarly flushes both queues, merging pendingQueue before draining.

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    while (changed) {
+      changed = false;
+      for (const add of allAdds) {
+        if (!transientIds.has(add.node.id) && transientIds.has(add.parentId)) {
+          transientIds.add(add.node.id);
+          changed = true;
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts:107]: Fixed in 25e54a8. The cascade was iterating over allAdds and triggering on any historical parentId, including ones from before a move. Fix: added lastParentById (built alongside lastAddEventIndex with last-write-wins semantics) and replaced the allAdds cascade iteration with a lastParentById iteration. Now only the node's final parent is checked — a child moved from a transient parent to a non-transient parent is correctly preserved. Added a regression test: child C added under transient P, then P removed and C moved to Q; result is P elided, C survives at Q.

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    for (const remove of data.removes) {
+      lastRemoveEventIndex.set(remove.id, i);
+    }
+  });
+
+  const transientIds = new Set<number>();
+  for (const [id, addIdx] of lastAddEventIndex) {
+    const removeIdx = lastRemoveEventIndex.get(id);
+    if (removeIdx !== undefined && addIdx < removeIdx) transientIds.add(id);
+  }
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts:63]: Fixed in fb1f7951. Added `firstAddEventIndex` and `firstRemoveEventIndex` maps. A node is now only marked transient when `firstAddIdx < firstRemoveIdx` (node was created within the window), guarding against the remove-add-remove case where the node pre-existed and `firstRemoveIdx` precedes `firstAddIdx`. Added a regression test.

File: packages/session-replay-browser/src/events/merge-mutation-events.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts:107]: Fixed in 25e54a8f. Switched the cascade from iterating `allAdds` (which includes historical parentIds before a move) to iterating `lastParentById` (only the final resting parent). A child moved away from a transient parent to a non-transient one is no longer elided.

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  const allAttributes = events.flatMap((e) => (e.data as mutationData).attributes);
+
+  const merged: mutationData = {
+    source: IncrementalSource.Mutation,
+    removes: transientIds.size > 0 ? allRemoves.filter((r) => !transientIds.has(r.id)) : allRemoves,
+    adds: transientIds.size > 0 ? allAdds.filter((a) => !transientIds.has(a.node.id)) : allAdds,
+    texts: transientIds.size > 0 ? allTexts.filter((t) => !transientIds.has(t.id)) : allTexts,
+    attributes: transientIds.size > 0 ? allAttributes.filter((a) => !transientIds.has(a.id)) : allAttributes,
+  };
+  return { ...first, data: merged } as eventWithTime;
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts:160]: Fixed in 0e4270f9. Introduced a 'pre-existing transient' category: when `firstRemoveIdx < firstAddIdx && lastAddIdx < lastRemoveIdx`, the re-add and all post-add removes are cancelled; only pre-add removes (the legitimate original removal) survive. This prevents the replayer's removes-before-adds ordering from re-creating the node. Cascaded to children whose final parent is a pre-existing-transient node. Added 6 regression tests covering the basic case, texts/attributes filtering, cascade, and the DR #PR update to the existing test to verify correct semantics.

**Comment by lewgordon-amplitude** [comment]
## PR Review Guide

**Summary:** Adds `mergeMutationEvents` — a utility that coalesces consecutive rrweb `IncrementalSnapshot` mutation events into one — and wires it into `EventCompressor`'s idle-callback queue and full-snapshot drain path. The goal is to reduce total stored events without changing replay semantics, by cancelling transient DOM nodes (added then removed within the same batch) before serialization.

**Risk:** Medium | 4 files | +635 -5 | Tests: passing (757 tests, 100% coverage)

### Dependency Graph

```
EventCompressor (event-compressor.ts)
  ├── pendingQueue  ──(on processQueue)──> mergeMutationTasks()
  │                                             └──> mergeMutationEvents()
  └── pre-snapshot drain ─────────────> mergeMutationTasks()
                                              └──> mergeMutationEvents()
                                                        └── mergeGroup()
                                                              ├── transientIds        (pure transient elision)
                                                              └── preExistingTransientIds  (re-add cancel)
```

### Review Priority

| # | File | Risk | What Changed | Review Tip |
|---|------|------|-------------|------------|
| 1 | `src/events/merge-mutation-events.ts` | Medium | New merge utility with transient elision | Verify all 4 remove-classification scenarios (pure transient, pre-existing transient, DOM move, add-then-move) against the rrweb removes-before-adds ordering guarantee |
| 2 | `src/events/event-compressor.ts` | Medium | Split queue into pending/task; add mergeMutationTasks | Confirm the pending→task split prevents re-merging; check the full-snapshot drain flushes both queues in the right order |
| 3 | `test/merge-mutation-events.test.ts` | Low | 22 unit tests for all edge cases | Verify the test names match the scenarios you'd worry about in production |
| 4 | `test/event-compressor.test.ts` | Low | 2 new processQueue tests for pending queue path | Spot-check the multi-session separation test |

### Test Gap Analysis
- OK `mergeMutationEvents` — 22 tests cover pure transient, pre-existing transient, DOM move, add-then-move, cascade, texts/attributes filtering, isAttachIframe isolation, mixed sequences
- OK `EventCompressor.mergeMutationTasks` — covered via `processQueue` tests pushing to `pendingQueue`
- OK Multi-session task separation — covered by the new "keeps tasks from different sessions separate" test
- ! No integration test verifying end-to-end: real rrweb events recorded → merged → replayed without "node not found" errors. The PR description flags this as a remaining manual validation item.
- ! The merge window size is unbounded (all events in the idle queue when `processQueue` fires). Under sustained high-mutation load this could be very large. No stress test or size cap.

### Key Observations
- **rrweb ordering invariant is the critical assumption.** The replayer always applies `removes` before `adds` within a single mutation event. Every piece of transient-elision logic exists to compensate for this: if a transient add+remove were kept, the remove would fire first (no-op), then the add would re-create the node. Reviewers should confirm this ordering guarantee holds in the rrweb version being used.
- **Four classification buckets, not two.** `mergeGroup` identifies: (1) pure transients (add then remove — cancel both), (2) pre-existing transients (remove, re-add, remove — cancel re-add and post-add removes), (3) DOM moves (remove then re-add — keep both), (4) add-then-move (add, then remove+re-add — keep at new parent). The `firstAddIdx === lastAddIdx === lastRemoveIdx` equality check for case 4 is particularly subtle.
- **The pending/task queue split.** Prior to this PR, all events entered `taskQueue` directly. Events in `taskQueue` may have already been merged; re-merging them corrupts move semantics for nodes that appear in multiple passes (a node moved in pass 1 would be incorrectly elided in pass 2). The split ensures only fresh `pendingQueue` events are ever passed to `mergeMutationTasks`.
- **isAttachIframe is never merged.** These events deliver a whole iframe document tree and must stay isolated. The guard is in `isMergeableMutation` and the PR has a dedicated test for it.

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
+import { EventType, IncrementalSource } from '@amplitude/rrweb-types';
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/events/merge-mutation-events.ts:1]: **[pr-copilot]** Medium risk

**What changed:** New file implementing `mergeMutationEvents` — coalesces consecutive mutation `IncrementalSnapshot` events into one, with transient-node elision to preserve replay correctness.
**Why it matters:** The merge logic sits directly in the serialization path for all session replay events. Incorrect elision produces phantom or missing DOM nodes in playback. There are four distinct cases (pure transient, pre-existing transient, DOM move, add-then-move) and all must be classified correctly.
**How to review:** Focus on `mergeGroup` lines 10–116. Trace the four classification branches starting at L53. Verify the `filteredRemoves` loop (L89–95) correctly splits pre-window removes from post-add removes. Confirm `lastParentById` cascade (L66–77) doesn't elide nodes that moved *away* from a cancelled parent.

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+  //                          location).
+  const transientIds = new Set<number>();
+  const preExistingTransientIds = new Set<number>();
+
+  for (const [id, firstAddIdx] of firstAddEventIndex) {
+    const firstRemoveIdx = firstRemoveEventIndex.get(id);
+    if (firstRemoveIdx === undefined) continue;
+    const lastAddIdx = lastAddEventIndex.get(id)!;
+    const lastRemoveIdx = lastRemoveEventIndex.get(id)!;
+    if (lastAddIdx >= lastRemoveIdx) continue; // ultimately present — keep as-is
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/events/merge-mutation-events.ts:53]: **[pr-copilot] !** Four cases hinge on this single `continue`. When `lastAddIdx >= lastRemoveIdx` the node is ultimately present — all three non-transient cases (DOM move, add-then-move, add-only) fall through here. Only nodes that are ultimately removed proceed to classification. Verify that `lastAddIdx == lastRemoveIdx` (add-then-move: add in e0, remove+re-add in e1) is correctly handled as "ultimately present".

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+  // Build filtered removes by iterating per event so we know each remove's event index.
+  // Pure transients:          drop all removes.
+  // Pre-existing transients:  drop removes at eventIdx >= firstAddIdx (the cancelled re-add cycle);
+  //                           keep removes at eventIdx < firstAddIdx (legitimate pre-window removal).
+  const filteredRemoves: mutationData['removes'][0][] = [];
+  events.forEach((e, eventIdx) => {
+    for (const r of (e.data as mutationData).removes) {
+      if (transientIds.has(r.id)) continue;
+      if (preExistingTransientIds.has(r.id) && eventIdx >= firstAddEventIndex.get(r.id)!) continue;
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/events/merge-mutation-events.ts:122]: **[pr-copilot] !** Pre-existing transient filtering: removes at `eventIdx >= firstAddIdx` are dropped (they're part of the cancelled re-add cycle); removes before `firstAddIdx` are kept (the legitimate original removal). Confirm that `firstAddEventIndex.get(r.id)!` is always set when a node is in `preExistingTransientIds` — it is, because the outer loop iterates `firstAddEventIndex`, but worth a quick mental check.

File: packages/session-replay-browser/src/events/event-compressor.ts [hunk 12]
```diff
@@ -3,6 +3,7 @@ import { EventType as RRWebEventType } from '@amplitude/rrweb-types';
 import type { eventWithTime } from '@amplitude/rrweb-types';
 import { SessionReplayJoinedConfig } from '../config/types';
 import { SessionReplayEventsManager } from '../typings/session-replay';
+import { mergeMutationEvents } from './merge-mutation-events';
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/events/event-compressor.ts:6]: **[pr-copilot]** Medium risk

**What changed:** Adds `pendingQueue` field; incoming events now queue in `pendingQueue` rather than `taskQueue` directly; `processQueue` merges the pending batch into `taskQueue` each call; full-snapshot drain flushes both queues.
**Why it matters:** The split between `pendingQueue` (unmerged) and `taskQueue` (already-merged) is critical: re-merging an already-merged event can corrupt move semantics for nodes that appear across multiple merge passes.
**How to review:** Check L93–99 (full-snapshot drain) — `taskQueue.splice(0)` is called *before* `mergeMutationTasks(pendingQueue.splice(0))`, ensuring pre-existing merged tasks are not re-merged. Check L122–123 (normal processQueue path) for the same pattern.

File: packages/session-replay-browser/src/events/event-compressor.ts [hunk 13]
```diff
@@ -88,8 +90,9 @@ export class EventCompressor {
       // Those events reference the pre-snapshot DOM and must be sent before
       // the full snapshot; if we let them be processed later they'd arrive at
       // the server after the snapshot and cause "node not found" replay errors.
-      if (this.taskQueue.length > 0) {
-        for (const task of this.taskQueue.splice(0)) {
+      if (this.taskQueue.length > 0 || this.pendingQueue.length > 0) {
+        const allTasks = [...this.taskQueue.splice(0), ...this.mergeMutationTasks(this.pendingQueue.splice(0))];
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/events/event-compressor.ts:94]: **[pr-copilot] !** Full-snapshot drain ordering: `taskQueue.splice(0)` (already-merged events) is concatenated with `mergeMutationTasks(pendingQueue.splice(0))` (freshly merged). Verify that already-merged tasks are emitted *first* — they represent events that arrived before the current `pendingQueue` batch, so they must precede the full snapshot.

File: packages/session-replay-browser/test/merge-mutation-events.test.ts [hunk 14]
```diff
@@ -0,0 +1,400 @@
+import { EventType, IncrementalSource } from '@amplitude/rrweb-types';
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/test/merge-mutation-events.test.ts:1]: **[pr-copilot]** Low risk — 22-test suite covering all four classification cases, cascade, texts/attributes filtering, and isAttachIframe isolation. 100% branch coverage.

File: packages/session-replay-browser/test/event-compressor.test.ts [hunk 15]
```diff
@@ -3,7 +3,7 @@ import { SessionReplayLocalConfig } from '../src/config/local-config';
 import { EventCompressor } from '../src/events/event-compressor';
 import { createEventsManager } from '../src/events/events-manager';
 import { SessionReplayEventsManager } from '../src/typings/session-replay';
-import { eventWithTime } from '@amplitude/rrweb-types';
+import { EventType, IncrementalSource, eventWithTime, mutationData } from '@amplitude/rrweb-types';
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/test/event-compressor.test.ts:6]: **[pr-copilot]** Low risk — adds 2 tests exercising the `pendingQueue → mergeMutationTasks → processQueue` path and multi-session separation.

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    const lastAddIdx = lastAddEventIndex.get(id)!;
+    const lastRemoveIdx = lastRemoveEventIndex.get(id)!;
+    if (lastAddIdx >= lastRemoveIdx) continue; // ultimately present — keep as-is
+
+    if (firstAddIdx < firstRemoveIdx) {
+      transientIds.add(id);
+    } else {
+      // firstRemoveIdx < firstAddIdx: pre-existing node removed, re-added, then removed again
+      preExistingTransientIds.add(id);
+    }
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts:60]: Fixed in 558fc578. Changed `else` to `else if (firstRemoveIdx < firstAddIdx)` so that `firstAddIdx === firstRemoveIdx` (same-event move) falls through to neither set. Added a regression test: same-event remove+add followed by a later remove now correctly preserves all three operations.

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+    let changed = true;
+    while (changed) {
+      changed = false;
+      for (const [nodeId, parentId] of lastParentById) {
+        if (
+          !transientIds.has(nodeId) &&
+          !preExistingTransientIds.has(nodeId) &&
+          (transientIds.has(parentId) || preExistingTransientIds.has(parentId))
+        ) {
+          transientIds.add(nodeId);
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts]: Fixed in commit 45613651. The cascade now checks whether the cascaded child itself is pre-existing (firstRemoveIdx < firstAddIdx) and routes it to preExistingTransientIds rather than transientIds, so its original pre-add removes are preserved. Added a regression test covering this exact scenario.

**Comment by lewgordon-amplitude** [comment]
bugbot run

**Review by stevenchien-amplitude — APPROVED** [review]
this looks good. just to clarify: the risk is for example instead of 20 mutations happening in 500ms, it'll just have 1 event that contains all 20 changes, so it's instantaneous instead of having timestamps for each specific mutation right? so if i'm scrubbing, it'll look like all 20 happened at once essentially.

**Comment by jxiwang** [comment]
Do we want to release this with a config users enable? I think right now it will be default on for everyone once released.

**Comment by lewgordon-amplitude** [comment]
@stevenchien-amplitude Yes, exactly right — merged events all carry the timestamp of the first event in the window, so scrubbing to a point mid-window would show all those mutations already applied. That's the core trade-off: fewer stored events at the cost of per-mutation timestamp granularity within the merge window.

Based on Jesse's feedback I've just added a `performanceConfig.mergeMutations` flag (default `false`) so the feature ships as opt-in. Until enabled, EventCompressor passes events through with no merging, so the replay experience is unchanged from today.

**Comment by lewgordon-amplitude** [comment]
@jxiwang Good call — just added `performanceConfig.mergeMutations` (default `false`) in 8dbb3bed. The feature is now opt-in; existing users see no change until they explicitly enable it.

**Review by jxiwang — APPROVED** [review]

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+          const nodeFirstAddIdx = firstAddEventIndex.get(nodeId);
+          if (
+            nodeFirstRemoveIdx !== undefined &&
+            nodeFirstAddIdx !== undefined &&
+            nodeFirstRemoveIdx < nodeFirstAddIdx
+          ) {
+            preExistingTransientIds.add(nodeId);
+          } else {
+            transientIds.add(nodeId);
+          }
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts:102]: Fixed in commit 41209358. Added a `cascadeDropAddsOnlyIds` set for the `nodeFirstRemoveIdx === nodeFirstAddIdx` case in the cascade: drops adds (going to the cancelled parent) while preserving removes from non-cancelled parents. The `else` branch no longer fires for same-event moves. Added regression tests covering this case, including a scenario where the node also has a subsequent remove from the cancelled parent (which is now correctly dropped).

File: packages/session-replay-browser/src/events/merge-mutation-events.ts [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+          changed = true;
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/events/merge-mutation-events.ts:107]: Fixed in commit 41209358. The root cause is the same as the sibling thread: when `nodeFirstRemoveIdx === nodeFirstAddIdx` (same-event move), the `else` branch in the cascade was incorrectly adding the node to `transientIds`, overriding the main loop's "ultimately present" classification and dropping its original-parent remove. The `cascadeDropAddsOnlyIds` fix handles this: drops adds to the cancelled parent while keeping removes from non-cancelled parents. Regression tests added.
